### PR TITLE
Add batch_predict

### PR DESCRIPTION
**Pull request by Ghallabi**
Add batch inference support

### CODE REVIEW

File: GroundingDINO/groundingdino/util/inference.py [hunk 1]
```diff
@@ -50,6 +50,47 @@ def load_image(image_path: str) -> Tuple[np.array, torch.Tensor]:
...
+    images = images.to(device)
+
+    with torch.no_grad():
+        outputs = model(images, captions=[caption]*images.shape[0])
+    
+    
+    tokenizer = model.tokenizer
+    prediction_logits = outputs["pred_logits"].sigmoid().cpu()
+    prediction_boxes = outputs["pred_boxes"].cpu()
+    tokenized = tokenizer(caption)
```
Comment (ldpalcu) [inline GroundingDINO/groundingdino/util/inference.py:74]: `tokenized = model.tokenizer(caption)` - one line

File: GroundingDINO/groundingdino/util/inference.py [hunk 2]
```diff
@@ -50,6 +50,47 @@ def load_image(image_path: str) -> Tuple[np.array, torch.Tensor]:
...
+    keep_mask = max_logits > box_threshold  # (B, nq)
+
+    logits_list = []
+    boxes_list = []
+    phrases_list = []
+
+    for i in range(prediction_logits.shape[0]):
+        logits_i = prediction_logits[i][keep_mask[i]]  # (n_i, 256)
+        boxes_i = prediction_boxes[i][keep_mask[i]]  # (n_i, 4)
+        phrases_i = [get_phrases_from_posmap(logit > text_threshold, tokenized, tokenizer).replace(".", "") for logit in logits_i]
```
Comment (ldpalcu) [inline GroundingDINO/groundingdino/util/inference.py:86]: Is it possible to rewrite `get_phrases_from_posmap` to accept a batch of logits?

File: GroundingDINO/groundingdino/util/inference.py [hunk 3]
```diff
@@ -50,6 +50,47 @@ def load_image(image_path: str) -> Tuple[np.array, torch.Tensor]:
...
+        box_threshold: float,
+        text_threshold: float,
+        device: str = "cuda"
+) -> Tuple[torch.Tensor, torch.Tensor, List[str]]:
+    caption = preprocess_caption(caption=caption)
+
+    model = model.to(device)
+    images = images.to(device)
+
+    with torch.no_grad():
```
[the remark this answers is not in the document]
Comment (Ghallabi) [reply GroundingDINO/groundingdino/util/inference.py:67]: The original codebase does this at model loading time 

```
def load_model(model_config_path: str, model_checkpoint_path: str, device: str = "cuda"):
    args = SLConfig.fromfile(model_config_path)
    args.device = device
    model = build_model(args)
    checkpoint = torch.load(model_checkpoint_path, map_location="cpu")
    model.load_state_dict(clean_state_dict(checkpoint["model"]), strict=False)
    model.eval()
    return model
 
```
    
 But good point worth adding it inside the predict function as well.

File: GroundingDINO/groundingdino/util/inference.py [hunk 4]
```diff
@@ -50,6 +50,47 @@ def load_image(image_path: str) -> Tuple[np.array, torch.Tensor]:
     return image, image_transformed
 
 
+
+def batch_predict(
```
Comment (calingc) [inline GroundingDINO/groundingdino/util/inference.py:54]: is there any build/setup related change that needs to happen?

If these are the only changes that need to happen in order to enable batching, and there's no need to change the setup script (avoiding cuda compilation for example), I see no reason why it couldn't be included in  `zia-vision`. It avoids having to keep dependencies in sync between the two different repos.

**Review by calingc — CHANGES_REQUESTED** [review]
The reason for forking the repo originally was to change the build script in order to skip cuda compilation for some parts of the code, and also modify existing inference code.

As far as I see in here, neither of these is needed

File: GroundingDINO/groundingdino/util/inference.py (hunk 4 above)
Comment (Ghallabi) [reply GroundingDINO/groundingdino/util/inference.py:54]: Short answer is no, I am also planning to modify the `forward` pass for `GroundingDino` model class. Which AFAIK will need to be rebuilt 🤔 - Am I correct @calingc ?

Comment (calingc) [reply GroundingDINO/groundingdino/util/inference.py:54]: So it's an existing implementation that has to be changed, it's not only the batch_predict above?

Comment (Ghallabi) [reply GroundingDINO/groundingdino/util/inference.py:54]: Yes, exactly!
